[GTK][WPE][Skia] Record all dirty tiles of a layer once, don't record each tile

nikolaszimmermann opened the pull request:
#### 5abe8ab78cbb3a00c4e190faa2b602ed7b199432
<pre>
[GTK][WPE][Skia] Record all dirty tiles of a layer once, don&apos;t record each tile
<a href="https://bugs.webkit.org/show_bug.cgi?id=290614">https://bugs.webkit.org/show_bug.cgi?id=290614</a>

Reviewed by Carlos Garcia Campos.

For each layer, we currently record each dirty tile, then replay in a worker
thread. We can do better and once record a SkPicture of the area spanning all
dirty tiles, and then use the same SkPicture for replaying N times, for each
dirty tile.

Covered by existing tests.

* Source/WebCore/platform/Skia.cmake:
* Source/WebCore/platform/SourcesSkia.txt:
* Source/WebCore/platform/graphics/skia/SkiaPaintingEngine.cpp:
(WebCore::SkiaPaintingEngine::paint):
(WebCore::SkiaPaintingEngine::record):
(WebCore::SkiaPaintingEngine::replay):
(WebCore::SkiaPaintingEngine::paintGraphicsLayerIntoBuffer const): Deleted.
(WebCore::SkiaPaintingEngine::paintLayer): Deleted.
(WebCore::SkiaPaintingEngine::postPaintingTask): Deleted.
(WebCore::SkiaPaintingEngine::performPaintingTask): Deleted.
* Source/WebCore/platform/graphics/skia/SkiaPaintingEngine.h:
(WebCore::SkiaPaintingEngine::useThreadedRendering const):
* Source/WebCore/platform/graphics/skia/SkiaRecordingResult.cpp: Added.
(WebCore::SkiaRecordingResult::SkiaRecordingResult):
(WebCore::SkiaRecordingResult::create):
(WebCore::SkiaRecordingResult::hasFences):
(WebCore::SkiaRecordingResult::waitForFenceIfNeeded):
* Source/WebCore/platform/graphics/skia/SkiaRecordingResult.h: Added.
* Source/WebCore/platform/graphics/skia/SkiaReplayCanvas.cpp:
(WebCore::SkiaReplayCanvas::SkiaReplayCanvas):
(WebCore::SkiaReplayCanvas::create):
(WebCore::SkiaReplayCanvas::waitForRenderingCompletionAndRewrapImageIfNeeded):
* Source/WebCore/platform/graphics/skia/SkiaReplayCanvas.h:
* Source/WebCore/platform/graphics/texmap/coordinated/CoordinatedBackingStoreProxy.cpp:
(WebCore::CoordinatedBackingStoreProxy::updateIfNeeded):
* Source/WebCore/platform/graphics/texmap/coordinated/CoordinatedPlatformLayer.cpp:
(WebCore::CoordinatedPlatformLayer::paint):
(WebCore::CoordinatedPlatformLayer::record):
(WebCore::CoordinatedPlatformLayer::replay):
* Source/WebCore/platform/graphics/texmap/coordinated/CoordinatedPlatformLayer.h:

Canonical link: <a href="https://commits.webkit.org/292929@main">https://commits.webkit.org/292929@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9bef1631cb8730c71e38cbea2743d7dd25bd7fd3

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/97497 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/17122 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/7338 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/102586 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/48026 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/99542 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/17415 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/25574 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/74294 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/31474 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/100500 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/13190 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/88171 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/54639 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/12959 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/6053 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/47468 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/82988 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/6134 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/104604 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/24576 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/17929 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/83340 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/24948 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/84297 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/82761 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/20819 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/27303 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/4991 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/18167 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/24538 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/29707 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/24360 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/27674 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/25934 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/2b2332f113d578d866ebebe2631819b262dbbfca. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/43183

carlosgcampos commented on Source/WebCore/platform/graphics/texmap/coordinated/CoordinatedBackingStoreProxy.cpp:
```diff
@@ -148,14 +153,25 @@ OptionSet<CoordinatedBackingStoreProxy::UpdateResult> CoordinatedBackingStorePro
 
+#if USE(SKIA)
+    // Record only once the whole layer.
+    std::optional<Ref<SkiaRecordingResult>> recording;
```
I would use RefPtr instead of optional Ref

nikolaszimmermann replied:
Fair enough.

carlosgcampos commented on Source/WebCore/platform/graphics/texmap/coordinated/CoordinatedBackingStoreProxy.cpp:
```diff
@@ -148,14 +153,25 @@ OptionSet<CoordinatedBackingStoreProxy::UpdateResult> CoordinatedBackingStorePro
+#if USE(SKIA)
+    // Record only once the whole layer.
+    std::optional<Ref<SkiaRecordingResult>> recording;
+    if (LIKELY(dirtyTilesCount > 0 && layer.client().paintingEngine().useThreadedRendering()))
```
I don't like that the backing store proxy needs to know about threaded rendering, but don't know a better solution. The other option would be that the paint engine knows about tiles, and that's not better.

nikolaszimmermann replied:
I came to the same conclusion, it's not pretty either way. At least it's minimal invasive this way.

carlosgcampos commented on Source/WebCore/platform/graphics/skia/SkiaRecordingResult.h:
```diff
@@ -0,0 +1,71 @@
+    bool hasFences() const { return !m_imageToFenceMap.isEmpty(); }
+    const sk_sp<SkPicture>& picture() const { return m_picture; }
+    IntRect recordRect() const { return m_recordRect; }
+    RenderingMode renderingMode() const { return m_renderingMode; }
```
Is this actually used?

nikolaszimmermann replied:
It has to be -- otherwise we can end up deciding between two different RenderingModes when we call determineRenderingMode() once in record() and once in replay() -- instead we should decide once on a RenderingMode, and stick with that during replay.

carlosgcampos reviewed (commented):
Awesome!

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/3ebbb60005f3540bd214015112d824c7440e8a93. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/43183

nikolaszimmermann commented on Source/WebCore/platform/graphics/skia/SkiaRecordingResult.h:
```diff
@@ -0,0 +1,71 @@
+
+    bool hasFences() const { return !m_imageToFenceMap.isEmpty(); }
+    const sk_sp<SkPicture>& picture() const { return m_picture; }
+    IntRect recordRect() const { return m_recordRect; }
```
Ok. I usually don't do that for almost-POD types, but it's fine. will adapt.

nikolaszimmermann commented on Source/WebCore/platform/graphics/skia/SkiaRecordingResult.h:
```diff
@@ -0,0 +1,71 @@
+
+    void waitForFenceIfNeeded(const SkImage*);
+
+    bool hasFences() const { return !m_imageToFenceMap.isEmpty(); }
```
Damn, indeed, that's missing. Will add.

nikolaszimmermann commented on Source/WebCore/platform/graphics/skia/SkiaRecordingResult.cpp:
```diff
@@ -0,0 +1,62 @@
+{
+    ASSERT(image);
+
+    if (auto fence = m_imageToFenceMap.take(image))
+        fence->serverWait();
```
Locking is fixed -- let's try EWS with server wait.

nikolaszimmermann commented on Source/WebCore/platform/graphics/skia/SkiaRecordingResult.h:
```diff
@@ -0,0 +1,71 @@
+
+class SkiaRecordingResult final : public ThreadSafeRefCounted<SkiaRecordingResult> {
+    WTF_MAKE_TZONE_ALLOCATED(SkiaRecordingResult);
+    WTF_MAKE_NONCOPYABLE(SkiaRecordingResult);
```
Ok. We did the same for SkiaReplayCanvas, will fix it as well.

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/fe836c4e80b723075eb72601e532ca0bd5a1360c. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/43183

nikolaszimmermann commented:
@carlosgcampos - do you know if we do run tests in HiDPI mode? Since I don't have a laptop with such a screen, could you give this a local try with Gtk+device scale? I couldn't verify if I'm taking into account the `contentsScale` correctly (eventually the simple `paintRect.location() - recordRect.location()` translation is too simple.

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/d683be13f365f0e632f8a9d1dfa66991b3e44cd0. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/43183

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/9bef1631cb8730c71e38cbea2743d7dd25bd7fd3. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/43183

nikolaszimmermann committed:
[GTK][WPE][Skia] Record all dirty tiles of a layer once, don't record each tile
https://bugs.webkit.org/show_bug.cgi?id=290614

Reviewed by Carlos Garcia Campos.

For each layer, we currently record each dirty tile, then replay in a worker
thread. We can do better and once record a SkPicture of the area spanning all
dirty tiles, and then use the same SkPicture for replaying N times, for each
dirty tile.

Covered by existing tests.

* Source/WebCore/platform/Skia.cmake:
* Source/WebCore/platform/SourcesSkia.txt:
* Source/WebCore/platform/graphics/skia/SkiaPaintingEngine.cpp:
(WebCore::SkiaPaintingEngine::paint):
(WebCore::SkiaPaintingEngine::record):
(WebCore::SkiaPaintingEngine::replay):
(WebCore::SkiaPaintingEngine::paintGraphicsLayerIntoBuffer const): Deleted.
(WebCore::SkiaPaintingEngine::paintLayer): Deleted.
(WebCore::SkiaPaintingEngine::postPaintingTask): Deleted.
(WebCore::SkiaPaintingEngine::performPaintingTask): Deleted.
* Source/WebCore/platform/graphics/skia/SkiaPaintingEngine.h:
(WebCore::SkiaPaintingEngine::useThreadedRendering const):
* Source/WebCore/platform/graphics/skia/SkiaRecordingResult.cpp: Added.
(WebCore::SkiaRecordingResult::SkiaRecordingResult):
(WebCore::SkiaRecordingResult::create):
(WebCore::SkiaRecordingResult::hasFences):
(WebCore::SkiaRecordingResult::waitForFenceIfNeeded):
* Source/WebCore/platform/graphics/skia/SkiaRecordingResult.h: Added.
* Source/WebCore/platform/graphics/skia/SkiaReplayCanvas.cpp:
(WebCore::SkiaReplayCanvas::SkiaReplayCanvas):
(WebCore::SkiaReplayCanvas::create):
(WebCore::SkiaReplayCanvas::waitForRenderingCompletionAndRewrapImageIfNeeded):
* Source/WebCore/platform/graphics/skia/SkiaReplayCanvas.h:
* Source/WebCore/platform/graphics/texmap/coordinated/CoordinatedBackingStoreProxy.cpp:
(WebCore::CoordinatedBackingStoreProxy::updateIfNeeded):
* Source/WebCore/platform/graphics/texmap/coordinated/CoordinatedPlatformLayer.cpp:
(WebCore::CoordinatedPlatformLayer::paint):
(WebCore::CoordinatedPlatformLayer::record):
(WebCore::CoordinatedPlatformLayer::replay):
* Source/WebCore/platform/graphics/texmap/coordinated/CoordinatedPlatformLayer.h:

Canonical link: https://commits.webkit.org/292929@main

webkit-commit-queue commented:
Committed 292929@main (5abe8ab78cbb): <https://commits.webkit.org/292929@main>

Reviewed commits have been landed. Closing PR #43183 and removing active labels.
